feat(linear_algebra): The Special linear group SL(n, R)

Vierkantor opened the pull request:
Definitions for the special linear group SL(n, R) (as `matrix.special_linear_group n R`), its group structure and the embedding SL(n, R) → GL(n, R) (as `matrix.special_linear_group.to_GL`).

Although this PR makes use of `matrix.det_mul`, whose type is changed in #1959, the two PRs should be able to be merged in either order since it is only invoked up to definitional equality.

TO CONTRIBUTORS:

Make sure you have:

  * [x] reviewed and applied the coding style: [coding](https://github.com/leanprover/mathlib/blob/master/docs/contribute/style.md), [naming](https://github.com/leanprover/mathlib/blob/master/docs/contribute/naming.md)
  * [x] reviewed and applied [the documentation requirements](https://github.com/leanprover/mathlib/blob/master/docs/contribute/doc.md)
  * [x] make sure definitions and lemmas are put in the right files
  * [x] make sure definitions and lemmas are not redundant

If this PR is related to a discussion on Zulip, please include a link in the discussion.

For reviewers: [code review check list](https://github.com/leanprover/mathlib/blob/master/docs/contribute/code-review.md)

Vierkantor committed:
Define the special linear group

Vierkantor committed:
Make definitions independent of PR #1959

That PR changes `det_mul` to have another, still definitionally equal, type.
If the invocations to `det_mul` are independent of syntactic equality, i.e. we
only pass `det_mul` to `erw`, this branch should be compatible with the state
before the change and after.

Vierkantor committed:
Documentation and code style improvements

Vierkantor committed:
Improve module docstring

jcommelin commented on src/linear_algebra/nonsingular_inverse.lean:
```diff
@@ -274,6 +277,11 @@ calc adjugate A ⬝ A = (Aᵀ ⬝ (adjugate Aᵀ))ᵀ :
   by rw [←adjugate_transpose, ←transpose_mul, transpose_transpose]
 ... = A.det • 1 : by rw [mul_adjugate (Aᵀ), det_transpose, transpose_smul, transpose_one]
 
+lemma det_adjugate_eq_one {A : matrix n n α} (h : A.det = 1) : (adjugate A).det = 1 :=
```
How hard would it be to show the general formula `det (adjugate A) = det(A)^{n.card - 1}` (assuming, maybe `n.card ≥ 1`)?

Vierkantor replied:
I managed to prove the general formula for the case where `det A` is invertible. The case `det A = 0` requires showing for all `B` that `det B = 0` iff there is a nonzero `v` such that `mul_vec B v = 0`, which seems to be quite a lot of work.

jcommelin replied:
Hmm, I'm not even sure if the general formula is true over a general ring... so assuming `det A` invertible is fine for now. It can always be refactored and generalised (-;

jcommelin commented on src/linear_algebra/special_linear_group.lean:
```diff
@@ -0,0 +1,133 @@
+
+We define the elements of `special_linear_group` to be matrices, since we need to
+compute their determinant. This is in contrast with `general_linear_group R M`,
+which consists of invertible `R`-linear maps on `M`.
```
Should we put this `special_linear_group` in the `matrix` namespace? We could also put another `general_linear_group` in the `matrix` namespace. Idem dito for orthogonal groups etc.
Then we can reserve `special_linear_group R V` for the group of determinant 1 linear endomorphisms.

Vierkantor replied:
These definitions already live in the `matrix` namespace :)

jcommelin replied:
Aah, sorry. I missed that.

sgouezel commented on src/linear_algebra/special_linear_group.lean:
```diff
@@ -0,0 +1,133 @@
+
+## Main definitions
+
+ * `matrix.special_linear_group` is the set of matrices with determinant 1
```
```suggestion
 * `matrix.special_linear_group` is the type of matrices with determinant 1
```
Also, you could maybe add here the parameters that one should feed to this function, i.e., `n R`.

sgouezel commented on src/linear_algebra/special_linear_group.lean:
```diff
@@ -0,0 +1,133 @@
+
+variables {n : Type u} [fintype n] [decidable_eq n] {R : Type v} [comm_ring R]
+
+lemma ext_iff (A B : special_linear_group n R) : A = B ↔ (∀ i j, A.1 i j = B.1 i j) :=
```
Could you register the coercion to `matrix n n R`, to be able to write `A i j` and so on?

jcommelin commented on src/linear_algebra/special_linear_group.lean:
```diff
@@ -0,0 +1,133 @@
+  ..special_linear_group.has_mul,
+  ..special_linear_group.has_one,
+  ..special_linear_group.has_inv,
+}
```
This brace should go to the end of the previous line.

jcommelin commented on src/linear_algebra/special_linear_group.lean:
```diff
@@ -0,0 +1,133 @@
+  ..A.val.to_lin
+}
+
+instance coe_GL : has_coe (special_linear_group n R) (general_linear_group R (n → R)) :=
```
I think I don't really like this coe. I would rather have a coe to matrices, and maybe to the `matrix.general_linear_group`. Of course, the latter doesn't exist yet, but maybe it should.
We probably need to think a bit about how to setup all the equivs between:
* matrices and linear endomorphisms
* `matrix.GL` and invertible lin. endo's
* `matrix.SL` and invertible lin endos with det = 1
* `matrix.O` and endomorphisms that preserve some symm bilinear form
* etc etc etc

this `coe` goes from the left column to the right column. Currently, that's because there's a bunch of holes in the definitions. I'm not sure if that justifies that it ends up in mathlib like this.

Vierkantor replied:
Makes sense. I'll add the coercion to matrices that @sgouezel also suggested and turn the `coe_GL` instance into a function like `to_linear_equiv`.

jcommelin reviewed (commented):
Thanks for working on this. My critique is not meant to be discouraging. I'm sorry that there are so many holes in the library...

Vierkantor committed:
Fix documentation

`matrix.special_linear_group` is not a set but a type

Co-Authored-By: sgouezel <sebastien.gouezel@univ-rennes1.fr>

Vierkantor committed:
Don't directly coerce from SL to linear maps

Now we coerce from `matrix.special_linear_group n R` to `matrix n n R` instead
of `general_linear_group R (n -> R)`

Vierkantor committed:
Whitespace fixes

Vierkantor committed:
Fix failing build in `src/linear_algebra/dual.lean`

Vierkantor committed:
Give an almost generic formula for `det_adjugate`

Vierkantor committed:
Move `det_eq_one_of_card_eq_zero` to the correct section

jcommelin commented on src/linear_algebra/nonsingular_inverse.lean:
```diff
@@ -274,6 +279,62 @@ calc adjugate A ⬝ A = (Aᵀ ⬝ (adjugate Aᵀ))ᵀ :
+    = A.det ^ (fintype.card n - 1) : det_adjugate_of_cancel (λ b hb, by simpa [h] using hb)
+... = 1                            : by rw [h, one_pow]
+
+lemma det_adjugate_of_invertible {A : matrix n n α} :
```
Maybe change the hypothesis to `is_unit A.det`? It shouldn't change the proof apart from the first line, and it would tie in to the API for units.

Vierkantor replied:
Ah, I should have realized that there was some nicer way to express that!

Vierkantor committed:
Replace the ad-hoc assumption of `det_adjugate_of_invertible` with `is_unit`

Vierkantor committed:
Fix linting error

There was an unnecessary assumption [decidable_eq α] floating around

sgouezel commented on src/linear_algebra/special_linear_group.lean:
```diff
@@ -0,0 +1,145 @@
+
+instance : inhabited (special_linear_group n R) := ⟨1⟩
+
+@[simp] lemma inv_val (A : special_linear_group n R) : A⁻¹.val = adjugate A.val := rfl
```
Once you have set up the coercion to matrices, the idea would be to avoid `val` altogether, and write the lemmas using the coercion (for instance writing the left hand side of this lemma as `(A⁻¹ : matrix n n R)`.

Vierkantor committed:
Replace `special_linear_group.val` with the appropriate coercions

sgouezel commented on src/data/matrix/basic.lean:
```diff
@@ -229,6 +229,14 @@ instance {β : Type w} [ring α] [add_comm_group β] [module α β] :
 variables [comm_ring α]
 
+lemma smul_eq_diagonal_mul [decidable_eq m] (M : matrix m n α) (a : α) :
+a • M = diagonal (λ _, a) ⬝ M :=
```
```suggestion
  a • M = diagonal (λ _, a) ⬝ M :=
```

sgouezel commented on src/linear_algebra/nonsingular_inverse.lean:
```diff
@@ -274,6 +280,62 @@ calc adjugate A ⬝ A = (Aᵀ ⬝ (adjugate Aᵀ))ᵀ :
   by rw [←adjugate_transpose, ←transpose_mul, transpose_transpose]
 ... = A.det • 1 : by rw [mul_adjugate (Aᵀ), det_transpose, transpose_smul, transpose_one]
 
+lemma det_adjugate_of_cancel {A : matrix n n α}
```
You could add a comment explaining what the lemma says and how you will use it.

sgouezel commented on src/linear_algebra/nonsingular_inverse.lean:
```diff
@@ -274,6 +280,62 @@ calc adjugate A ⬝ A = (Aᵀ ⬝ (adjugate Aᵀ))ᵀ :
+    = A.det ^ (fintype.card n - 1) : det_adjugate_of_cancel (λ b hb, by simpa [h] using hb)
+... = 1                            : by rw [h, one_pow]
+
+lemma det_adjugate_of_is_unit {A : matrix n n α} (h : is_unit A.det) :
```
Add a comment that the assumption `h` is not necessary, but that this would require a different proof.

sgouezel commented on src/linear_algebra/special_linear_group.lean:
```diff
@@ -0,0 +1,169 @@
+
+variables {n : Type u} [fintype n] [decidable_eq n] {R : Type v} [comm_ring R]
+
+instance coe_matrix : has_coe (special_linear_group n R) (matrix n n R) :=
+⟨λ A, A.val⟩
+
+instance coe_fun : has_coe_to_fun (special_linear_group n R) :=
+{ F   := λ _, n → n → R,
+  coe := λ A, A.val }
```
I am not an expert on coercions, but is it really necessary to have the two of them? It forces you to state two versions of several lemmas below. If you just register the coercion to matrices, does it work out of the box, inserting the coercion from matrices to functions when needed?

Vierkantor replied:
I couldn't get to work with only the `coe_matrix` instance, even with inserting explicit coercions. (Perhaps the reason is that `matrix n n R` is defined to be *equal* to `n → n → R`, instead of having a cast?)

Vierkantor committed:
whitespace

Correctly indent continued line

Co-Authored-By: sgouezel <sebastien.gouezel@univ-rennes1.fr>

Vierkantor committed:
Docstrings for the `det_adjugate_of_...` lemmas

mergify[bot] committed:
Merge branch 'master' into special_linear_group
